src/query.rs

```diff
@@ -508,6 +533,17 @@ pub struct QueryMut<'q, Q: Query> {
 
 impl<'q, Q: Query> QueryMut<'q, Q> {
     pub(crate) fn new(meta: &'q [EntityMeta], archetypes: &'q mut [Archetype]) -> Self {
+        // This looks like an ugly O(n^2) loop, but everything's constant after inlining, so in
+        // practice LLVM optimizes it out entirely.
+        Q::Fetch::for_each_borrow(|a, i, unique| {
+            if unique {
```

Comment:
One I think I did wonder is whether "properly" checking borrowing is necessary or whether just checking uniqueness of the component types would suffice. Borrowing the same component type twice immutably is a bit useless, but it is also possible using the dynamically checked variant. So it would be a strange API difference, but it would also simplify the implementation by avoid the uniqueness tag and the counter.

Reply:
How would we avoid the counter in that case?

Reply:
> it would also simplify the implementation by avoiding the uniqueness tag and the counter.

Actually, this would not be the case as we still need to avoid checking a component type against itself. But so we do need to resolve https://github.com/Ralith/hecs/pull/162#discussion_r628839322

Reply:
We cross posted here...